internal/datastore/proxy/context.go

```diff
@@ -0,0 +1,117 @@
+package proxy
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/authzed/spicedb/internal/datastore/options"
+	log "github.com/authzed/spicedb/internal/logging"
+	"github.com/authzed/spicedb/pkg/datastore"
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+)
+
+// SeparateContextWithTracing is a utility method which allows for severing the
+// context between grpc and the datastore to prevent context cancellation from
+// killing database connections that should otherwise go back to the connection
+// pool.
+func SeparateContextWithTracing(ctx context.Context) context.Context {
+	span := trace.SpanFromContext(ctx)
+	ctxWithObservability := trace.ContextWithSpan(context.Background(), span)
+
+	loggerFromContext := log.Ctx(ctx)
+	if loggerFromContext != nil {
+		ctxWithObservability = loggerFromContext.WithContext(ctxWithObservability)
+	}
+
+	return ctxWithObservability
+}
+
+// NewSeparatingContextDatastoreProxy severs any timeouts in the context being
+// passed to the datastore and only retains tracing metadata.
+//
+// This is useful for datastores that do not want to close connections when a
+// cancel or deadline occurs.
+func NewSeparatingContextDatastoreProxy(d datastore.Datastore) datastore.Datastore {
+	return &ctxProxy{d}
+}
+
+type ctxProxy struct{ delegate datastore.Datastore }
+
+func (p *ctxProxy) ReadWriteTx(ctx context.Context, f datastore.TxUserFunc) (datastore.Revision, error) {
+	return p.delegate.ReadWriteTx(ctx, f)
```

Comment:
maybe clarify we explicitly not want to sever context in operations that may involve writes?